src/components/auto-complete/auto-complete.tsx

```diff
@@ -1,131 +1,283 @@
-            this.openPopup();
+    public blur(): void {
+        this.input!.blur();
+    }
+
+    protected renderPopup() {
+        const popupContents = this.filteredItems.length ?
+            (
+                <SelectionListView
+                    className="list"
+                    focused
+                    list={this.list}
+                    onClick={this.handleListClick}
+                    onMouseDown={this.handleListMouseDown}
+                />
+            ) :
+            (
+                <div
+                    data-automation-id="NO_SUGGESTIONS_NOTICE"
+                    className="noSuggestionsNotice"
+                    onMouseDown={this.handleNoSuggestionsNoticeMouseDown}
+                >
+                    {this.noSuggestionsNotice}
+                </div>
+            );
+
+        return popupContents;
+    }
+
+    @action protected handleFocus = (): void => {
+        this.focused = true;
+        if (this.openOnFocus && this.isEditable) {
+            this.popupTriggered = true;
         }
     }
 
-    private onClick = (e: React.MouseEvent<HTMLElement>, itemIndex: number) => {
-        if (itemIndex > -1) {
-            this.props.onChange!({value: this.list.items[itemIndex].value});
+    @action protected handleBlur = (): void => {
+        this.focused = false;
+        this.popupTriggered = false;
+    }
+
+    @action protected handleInputChange: React.ChangeEventHandler<HTMLInputElement> = event => {
+        this.popupTriggered = true;
+        this.props.onChange!({value: this.input!.value});
+    }
+
+    @action protected handleInputKeydown: React.KeyboardEventHandler<HTMLInputElement> = event => {
+        const isPopupOpen = this.isPopupOpen;
+
+        switch (event.keyCode) {
+            case keycode('escape'):
+                this.popupTriggered = false;
+                this.list.focusIndex(-1);
+                break;
+
+            case keycode('enter'):
+                if (isPopupOpen && this.list.focusedIndex > -1) {
+                    event.preventDefault();
+                    this.popupTriggered = false;
+                    const itemIndex = this.list.focusedIndex;
+                    this.list.focusIndex(-1);
+                    this.props.onChange!({value: this.list.items[itemIndex].label});
+                }
+                break;
+
+            case keycode('down'):
+                if (isPopupOpen) {
+                    event.preventDefault();
+                    this.list.focusNext();
+                }
+                break;
+
+            case keycode('up'):
+                if (isPopupOpen) {
+                    event.preventDefault();
+                    this.list.focusPrevious();
+                }
+                break;
         }
-        this.togglePopup();
     }
 
-    private onCaretClick = () => {
-        this.togglePopup();
+    protected handleNoSuggestionsNoticeMouseDown: React.MouseEventHandler<HTMLElement> = event => {
```

Comment:
This isn't tested

Reply:
Good catch! Unfortunately there's no way to test this using simulated events. Simulated mouseDown doesn't steal focus from the input, so the test would pass regardless of whether this code works or not.

Reply:
I see.